lib/posix-futex: Add initial futex implementation

adinasm opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->
 - `CONFIG_LIBPOSIX_FUTEX=y`

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR introduces `posix-futex` which contains the implementation for the futex system call. At the moment, it supports the following futex operations:
- `FUTEX_WAIT`
- `FUTEX_WAKE`
- `FUTEX_CMP_REQUEUE`

It also contains some basic tests for the aforementioned operations.

danield20 committed:
lib/posix-futex: Library skeleton

posix-futex will encapsulate Linux-compatible futex system calls.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

danield20 committed:
lib/posix-futex: Initial implementation

This commit introduces the initial futex system call implementation. The
supported operations are FUTEX_WAKE, FUTEX_WAIT and FUTEX_CMP_REQUEUE.

Signed-off-by: Daniel Dinca <dincadaniel97@gmail.com>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>
Signed-off-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
Signed-off-by: Adina Smeu <adina.smeu@gmail.com>

aadhi0319 commented:
**Menuconfig:**
- posix-futex
- syscall-shim
- uktest

Running `qemu-guest -k app-helloworld_kvm-x86_64` causes a crash on `posix_futex_testsuite->test_cmp_requeue_two_waiters_no_requeue`. @adinasm is this reproducible on your end?

adinasm commented:
Hi @aadhi0319, I cannot reproduce this crash. Can you give me more details about it?

Sairajkodilkar commented:
Hey @adinasm thanks for the PR

- You should protect the `futex_list` using a spinlock because the list APIs `uk_list_add_tail()`, `uk_list_for_each_safe()` and `uk_list_del()` are not SMP safe.

- In `futex_wait()` function instead of iterating over the entire list, you can check if the thread is runnable or not.

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+};
+
+static UK_LIST_HEAD(futex_list);
+
```
uk_spinlock futex_list_lock = UK_SPINLOCK_INITIALIZER();

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+
+	if (ukarch_load_n(uaddr) == val) {
+
+		/* Enqueue thread to wait list */
```
irqf = uk_plat_save_irqf();
uk_spin_lock(&futex_list_lock);

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+
+		/* Enqueue thread to wait list */
+		uk_list_add_tail(&f.list_node, &futex_list);
+
```
uk_spin_unlock(&futex_list_lock);
uk_plat_restore_irqf(irqf);

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+
+		uk_sched_yield();
+
+		/* Thread is no longer blocked. Remove it from the futex list */
```
irqf = uk_plat_save_irqf();
uk_spin_lock(&futex_list_lock);

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+
+		/* Thread is no longer blocked. Remove it from the futex list */
+		uk_list_del(&f.list_node);
+
```
uk_spin_unlock(&futex_list_lock);
uk_plat_restore_irqf(irqf);

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+
+	if (val == 0)
+		return 0;
+
```
irqf = uk_plat_save_irqf();
uk_spin_lock(&futex_list_lock);

Sairajkodilkar commented on lib/posix-futex/futex.c:
```diff
@@ -0,0 +1,188 @@
+				break;
+		}
+	}
+
```
uk_spin_unlock(&futex_list_lock);
uk_plat_restore_irqf(irqf);

aadhi0319 commented on lib/posix-futex/tests/test_posix_futex.c:
```diff
@@ -0,0 +1,420 @@
+	threads[0] = uk_thread_create("Waiter 1", waiter_func, args + 0);
+	threads[1] = uk_thread_create("Waiter 2", waiter_func, args + 1);
+	threads[2] = uk_thread_create("Requeuer", requeuer_func, args + 2);
+
```
It seems that `threads[3]` is not initialized. I am getting a crash when `uk_thread_wait(threads[3])` is called in the for loop below.

adinasm replied:
Thanks, now I set the right number of threads.

aadhi0319 replied:
All tests pass for me now.

aadhi0319 commented on lib/posix-futex/Config.uk:
```diff
@@ -0,0 +1,19 @@
+	select LIBUKTEST
+	select LIBSYSCALL_SHIM
+
+endif
```
Is there a way to enable `LIBSYSCALL_SHIM` if this library is added and all tests are enabled from `uktest` in menuconfig? In that scenario, there is an error during make which may be confusing to users who are not familiar with unikraft.

adinasm replied:
I added another select in case all the tests are enabled from `uktest` and the error is gone now on my end.

aadhi0319 replied:
I believe this has the opposite effect. From my understanding, this selects the option to run all uktests if libfutex tests are enabled, which is not the functionality we desire. We may need to modify the Config.uk for uktest to get the intended behavior. @razvand would probably be the best point of contact to determine how to address this (including whether we even need to address this).

aadhi0319 commented on lib/posix-futex/tests/test_posix_futex.c:
```diff
@@ -0,0 +1,420 @@
+#include <uk/syscall.h>
+#include <uk/sched.h>
+
+/* TODO: Replace with the real futex system call number */
```
Has this TODO been implemented? If so, I would suggest removing this comment.

adinasm replied:
Removed it.

adinasm commented:
>     * In `futex_wait()` function instead of iterating over the entire list, you can check if the thread is runnable or not.

@Sairajkodilkar I'm not sure that the timed out thread should check if it's runnable. When `schedcoop_schedule()` is called, the expired threads are awaken [1]. This means that even though the thread timed out in `futex_wait()`, it would already be set as runnable by the time it resumes after `sched_yield()`. Please correct me if I'm wrong.

[1] https://github.com/unikraft/unikraft/blob/staging/lib/ukschedcoop/schedcoop.c#L87

aadhi0319 reviewed (changes requested):
Thanks for the PR. Looks pretty good from what I can tell; just made a couple small suggestions. Once that test is fixed I'll run the test cases again and make sure that the tests pass on my end too.

Sairajkodilkar commented:
Hi, @adinasm yeah, you are right about it! sorry I missed that part of the scheduler

Sairajkodilkar commented:
Hey @adinasm, the PR looks great, can you please resolve the checkpatch warnings ?

adinasm committed:
lib/posix-futex: Add initial tests

Added tests for the currently implemented futex operations (wait, wake and
cmp_requeue).

Signed-off-by: Adina Smeu <adina.smeu@gmail.com>

skuenzer commented:
This PR is okay from my side. I would approve as soon as review is finished.
Thanks,

Simon

marcrittinghaus commented:
I already use this PR successfully in a project that requires futexes and never had problems. So it seems to work quite well 😃👍🏻

@Sairajkodilkar By chance, do you have the time to check this out today?

Sairajkodilkar commented:
Hi @marcrittinghaus  the PR is okay from my side too.

razvand reviewed (approved):
All OK. Thanks, @adinasm 

Reviewed-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

skuenzer reviewed (approved):
Approved-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
